File: 008_String_To_Integer/solution.go
const (
    MAX int = 1<<31 - 1
    MIN int = -1 * MAX - 1
)

func myAtoi(str string) int {
    if len(str) == 0 {
        return 0
    }
    runes := []rune(str)
    sign := 1
    i := 0
    for ; runes[i] == rune(' '); i ++ {
    }
    
    if runes[i] == rune('-') {
        sign = -1
        i ++
    } else if runes[i] == rune('+') {
        i ++
    }
    
    ret := 0
    for ; i < len(str); i ++ {
        val := int(runes[i] - rune('0'))
        if val < 0 || val > 9 {
            break;
        } else {
            ret = 10 * ret + val
        }
        if ret * sign > MAX {
            return MAX
        }
        if ret * sign < MIN {
            return MIN
        }
    }
    return ret * sign
}

File: 008_String_To_Integer/solution.py
class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        str = str.strip()
        if not str:
            return 0
            
        MAX_INT = 2147483647
        MIN_INT = -2147483648
        ret = 0
        pos = 0
        overflow = False
        sign = 1
        
        if str[pos] == '-':
            sign = -1
            pos += 1
        elif str[pos] == '+':
            pos += 1
            
        for i in range(pos, len(str)):
            if not str[i].isdigit():
                break;
            ret = ret * 10 + int(str[i])
            if not MIN_INT <= sign * ret <= MAX_INT:
                overflow = True
                break
        if overflow:
            return MAX_INT if sign == 1 else MIN_INT
        else:
            return sign * ret
